lib/src/git.rs

```diff
@@ -546,13 +546,15 @@ fn remotely_pinned_commit_ids(view: &View) -> Vec<CommitId> {
 ///
 /// Unlike `reset_head()`, this function doesn't move the working-copy commit to
 /// the child of the new `HEAD@git` revision.
-pub fn import_head(mut_repo: &mut MutableRepo) -> Result<(), GitImportError> {
+pub fn import_head(
+    mut_repo: &mut MutableRepo,
+    git_worktree: &gix::Repository,
+) -> Result<(), GitImportError> {
```

Comment:
@martinvonz, Is it acceptable to teach `RepoLoader` about `workspace_root: Option<&Path>`? I know it's layering violation, but will simplify the mess of backend Git repo and workspace's Git repo.

The idea is roughly as follows:
```
RepoLoader::init_from_file_system(user_settings, &self.repo_path, Some(self.workspace_root()), store_factories)
  store_factories.load_backend(user_settings, &repo_path.join("store"), maybe_workspace_root)
    GitBackend::load(settings, store_path, maybe_workspace_root)
      # check if the workspace_root points to the Git repo of the store.
      # if it does, load it from there
```

Reply:
I would prefer not. I recently removed `RepoLoader::repo_path` because paths don't make sense when `RepoLoader` is used on a server (4fb7dba73df0bb0241a51d35f04d66f2ae450be6). I haven't started reviewing this PR yet so I don't know if there's another good solution.

Reply:
@martinvonz gentle ping. I expect the decision will affect fundamental design.

@cormacrelf do you like the idea of loading (or reinitializing) GitBackend for the workspace/worktree path (instead of working around the problem at CLI layer)?

Reply:
I think it would be good to have it in `jj_lib`. Currently it is CLI-only so other lib-only projects are not going to be able to use it. If `libgit2` supports worktrees, then so should `jj_lib` I think.

 I don't have an opinion on which is better, but it seems like two options are:

-  reinitialize in someplace like `DefaultWorkspaceLoader::load`
-  or have it pass down a `workspace_root: Option<&Path>` into the repo loader.

Reply:
I tried implementing this, easy until you get to `workspace add`. Even if you pass `workspace_root: Option<&Path>` down through RepoLoader, the code for adding a workspace needs to reinitialise an existing repo with a different workspace root. The backend is stored behind multiple Arc references and quite difficult to modify.

Reply:
> @martinvonz gentle ping. I expect the decision will affect fundamental design.

Sorry, I had not understood that you were expecting my input. I don't know what the issue here is, so I'll try to review this PR now.

Reply:
> I tried implementing this, easy until you get to `workspace add`. Even if you pass `workspace_root: Option<&Path>` down through RepoLoader, the code for adding a workspace needs to reinitialise an existing GitBackend with a different workspace root.

Good point. Maybe we can work around it by reloading workspace/repo pair, but it's yeah a bit ugly.

> > @martinvonz gentle ping. I expect the decision will affect fundamental design.
> 
> Sorry, I had not understood that you were expecting my input. I don't know what the issue here is, so I'll try to review this PR now.

It's basically whether we accept this PR (adding workarounds to CLI) or not (to GitBackend and its loaders.) For example, https://github.com/martinvonz/jj/pull/4644/commits/70dae549c85451c31d9f8a774c01636939f9191c wouldn't be needed if `GitBackend::git_workdir()` points to the jj's workspace root.

Reply:
Another idea:
* remove `GitBackend::git_workdir()`, etc.
* rename `GitBackend::git_repo()` to `::git_store_repo()`? (ideally it should be detached from the workdir)
* add `GitBackend::git_repo(workspace_root: &Path) -> Option_or_Result<_>`, etc. which check if the workdir points to the same repository, load+cache or reuse repository instance (which shares in-memory cache)

Reply:
The problem is that we have to be careful to not use `git_backend.git_repo()` when `HEAD` matters. IIRC, this patch re-instantiates git2/gix repo by CLI, and passes it to `git::import_head()`/`reset_head()`/`export_refs()` because `git_backend.git_repo()` points to the worktree of the main/default workspace.

If `GitBackend::load()` could set up the internal `gix::ThreadSafeRepository` properly, all problems would go away.

Reply:
I sketched out what having GitBackend open the right repo would look like in https://github.com/martinvonz/jj/pull/4801. The code supporting `jj workspace add`'s reinitialization needs a better idea than using the arc-swap crate, which I just whacked in there to make it go. But it is feasible.

Reply:
Thanks. I assume `is_colocated_git_workspace()` and `GitBackend` constructors will become much simpler if we remove colocation warning and refactor the code. I think it's okay to ignore non-colocated `.git` directory.

For `jj workspace add`, I would just make it reload the entire repo at the new workspace.

Reply:
Reloading RepoLoader from disk was easy. Neat!

I refactored the code but kept the colocation warnings by stuffing them in an enum. The GitBackend code ended up not getting messy at all.

Reply:
Thanks. I'll take a look tomorrow. btw, warning could be moved to CLI if that's important. Something like, if workspace_root + "/.git" exists, and if not git_repo.is_colocated(), emit warning.

Reply:
I generally like the new patches, many thanks.

@martinvonz, are you okay with this change?
https://github.com/martinvonz/jj/pull/4644/commits/8b586b0c54ac5bd84ff5e9f5e2871252d457e27d